Revert revert of downward api node defaults

derekwaynecarr opened the pull request:
Reverts the revert of https://github.com/kubernetes/kubernetes/pull/27439

Fixes #27062

@dchen1107 - who at Google can help debug why this caused issues with GKE infrastructure but not GCE merge queue?

/cc @wojtek-t @piosz @fgrzadkowski @eparis @pmorie

derekwaynecarr commented:
cc @alex-mohr @fabioy @roberthbailey

dchen1107 commented:
Here is what I found so far, all worker nodes work fine without any issues except it cannot connect to API server. GKE's image is GCI image, not old, to-be-deprecated containerVM image. I suspect there is a misconfiguration on GCI master node. cc/ @andyzheng0831 @wonderfly can you guys help to debug this on GKE?

andyzheng0831 commented:
Does it only happens in GKE master but not k8s master? I can create a cluster to take a check

zmerlynn commented:
@dchen1107: I believe there's some odd issue here because GKE is sideloading into kubelet, and it's interacting with @derekwaynecarr's PR:

```
Jun 15 13:37:58 gke-249458424157-58f3136937cf94bde20b kubelet[878]: I0615 13:37:58.135895     878 kubelet.go:2650] SyncLoop (ADD, "http"): "<redacted>"
Jun 15 13:37:58 gke-249458424157-58f3136937cf94bde20b kubelet[878]: E0615 13:37:58.136338     878 pod_workers.go:183] Error syncing pod c88a360a9d57cab55ccd658fcea87956, skipping: failed to find node object in cache, expected a non-nil object in the cache.
```

That redacted pod was sideloaded through the `google-container-manifest` metadata, and it seems to be poisoning loading the master pods.

dchen1107 commented:
@andyzheng0831 Yes, only on GKE master which has different configuration from k8s master. It failed consistently last night for many runs. But I couldn't spot the root cause from the pr itself. 
@zmerlynn I think you might be right on this, but couldn't figure out the relationships among sideloading, GCI image and this pr. Need a debug cluster for this. :-)

andyzheng0831 commented:
I am merging this change into my local repo and create a GKE cluster to investigate. Will update you ASAP

andyzheng0831 commented:
@derekwaynecarr @dchen1107 double check with you: What kind of problem did you see in GKE? What I see is that cluster creation is stuck. After sshing into the master, it seems that api-server is not up. Is this the same as your cluster?

dchen1107 commented:
@andyzheng0831 That is the problem what I guessed here.

wojtek-t commented:
@andyzheng0831 - yes, this is exactly the problem that we observed.

dchen1107 commented:
@derekwaynecarr I think I know what is going on here: on GKE, master node is not registered to apiserver, not like GCE. We called it as standalone model. But in your pr, you tried to store node object in cache which triggers the problem

 @roberthbailey, we really should fix this unnecessary discrepancy between GKE and the rest of world.

zmerlynn commented:
@dchen1107: ... you mean by fixing it on GCE? :)

derekwaynecarr commented on pkg/kubelet/kubelet_resources.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	lastUpdatedNodeObject := kl.lastUpdatedNodeObject.Load()
+	if lastUpdatedNodeObject == nil {
```
@dchen1107 - so if I update this function to return if `kl.standaloneMode` is true, would that be sufficient for GKE?

dchen1107 replied:
I think that is sufficient for GKE, but I need to double check. I am in release burndown meeting, will review your patch after that. Thanks!

derekwaynecarr replied:
I will see if there is a cleaner option, maybe pull the capacity from cadvisor.MachineInfo if node is not present...

roberthbailey commented:
Yeah, we'd need to do some security hardening (among other changes) before we could change it on gke.

derekwaynecarr commented:
Well, this rebase is not so simple all of a sudden ;-)

derekwaynecarr commented:
@dchen1107 - the second commit updates the PR to fall back to cadvisor if the node was not registered with the api server.  if cadvisor is not available, it defaults the capacity for the downward api to 0.  Assuming this passes the PR checkers, can someone verify this branch on GKE?

aaronlevy commented:
Wondering if it might be generally useful to have the lastUpdatedNodeObject pre-seeded by initialNodeState() regardless of contacting an api server. Is there some reason to maintain this isolation (and having to special case stuff like this)?

For reference, I'd like to submit a fix for hostIP not being populated for static pods when api-server hasn't been contacted. A relatively easy fix is just pre-seeding a nodeCache with the initial node state. But could instead follow the pattern proposed in this PR too.

Ref: 
https://github.com/kubernetes/kubernetes/issues/26590
https://github.com/kubernetes/kubernetes/pull/26602

derekwaynecarr commented:
@aaronlevy - I see your point.  Let me take another pass at this.

derekwaynecarr commented:
@aaronlevy - updated the code to populate `lastUpdatedNodeObject` even when in standalone mode as part of initial node status call.  PTAL.

aaronlevy commented on pkg/kubelet/kubelet.go:
```diff
@@ -1900,7 +1919,11 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 
 	// Wait for volumes to attach/mount
-	if err := kl.volumeManager.WaitForAttachAndMount(pod); err != nil {
+	defaultedPod, _, err := kl.defaultPodLimitsForDownwardApi(pod, nil)
```
Wouldn't this error still in the "standalone" case? It would seem that `LastNodeUpdateObject` will still be empty as initialNodeStatus() is called from registerWithAPIServer()

derekwaynecarr replied:
It was called in getNode and getNodeAnyway

I don't like initial node status modifying Kubelet state.  I will move setting last node update object into Kubelet Run when setup to not register node based on the value returned from initial node status.

This change will have to wait until tomorrow.

derekwaynecarr replied:
We really should refactor some aspects of standalone mode handling in Kubelet.  It's kind of a hidden scenario in code base.

aaronlevy replied:
ah, right I see.
+1 on moving to Run()

derekwaynecarr replied:
commenting for future reviewer clarity, I removed 'lastUpdatedNode` altogether from the orginal version of this PR to keep change set minimal at this point in the release.

aaronlevy commented on pkg/kubelet/kubelet.go:
```diff
@@ -1152,6 +1163,10 @@ func (kl *Kubelet) registerWithApiserver() {
 		}
+
+		// Cache the node object.
+		kl.lastUpdatedNodeObject.Store(node)
```
Assuming the initial seed-state is moved to kubelet.Run, should this only be updated if Create() below is successful?

Just a side note, this starts to feel a little fragile - now anywhere someone happens to update the node, they also need to know to populate this field?

pmorie commented:
@derekwaynecarr @aaronlevy 

> We really should refactor some aspects of standalone mode handling in Kubelet. It's kind of a hidden scenario in code base.

Agree.  Separate commonalities of normal and standalone modes into a shared bootstrap function then gate to mode-specific initializer?

> Just a side note, this starts to feel a little fragile - now anywhere someone happens to update the node, they also need to know to populate this field?

Agree -- we need to ensure that this is updated wherever a write of the node occurs:

```
pkg/kubelet/kubelet.go:1171:            if _, err := kl.kubeClient.Core().Nodes().Create(node); err != nil {
pkg/kubelet/kubelet.go:3392:                    if updatedNode, err := kl.kubeClient.Core().Nodes().Update(node); err != nil {
pkg/kubelet/kubelet.go:3407:    _, err = kl.kubeClient.Core().Nodes().UpdateStatus(node)
```

We could extract method within kubelet for:

- `CreateNode`
- `UpdateNode`
- `UpdateNodeStatus`

...and use those to ensure that the reference is updated.

derekwaynecarr committed:
Downward API defaults resource limits to node capacity/allocatable

derekwaynecarr commented:
@pmorie @aaronlevy 

I removed the `lastUpdatedNode` concept from this PR altogether in favor of just calling `getNodeAnyWay` which works fine with either standalone or not standalone mode.  When not standalone, it will be the node from the api server, and when standalone, it will be derived from `initialNodeStatus` which also has the static capacity/allocatable values needed by this PR.

@dchen1107 - we should let this pass the PR checks, but if it does, it would be good if GKE can be tested against this PR prior to merge.

aaronlevy commented:
I like the getNodeAnyWay() approach... seems much simpler.

derekwaynecarr commented:
@andyzheng0831 - this appears to have passed all PR checkers, are you able to verify this branch will not break a GKE environment?

pmorie commented on pkg/kubelet/kubelet_resources.go:
```diff
@@ -0,0 +1,84 @@
+
+	node, err := kl.getNodeAnyWay()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed find node object, expected a node")
```
nit: failed to find

pmorie commented on pkg/kubelet/kubelet_resources.go:
```diff
@@ -0,0 +1,84 @@
+	}
+	outputPod, ok := podCopy.(*api.Pod)
+	if !ok {
+		return nil, nil, fmt.Errorf("unexpected type")
```
clarify what this message is in relation to?

pmorie commented:
@derekwaynecarr were you going to add an e2e for the volume plugin?

derekwaynecarr commented:
@pmorie - I can, can you link to a pointer on where I should write that test?  Is there a model I should follow?

erictune commented:
@derekwaynecarr @pmorie can the fix please go in now, to unblock the release, and the e2e can come in a later PR?

derekwaynecarr commented:
@erictune +1  - i would like this merged now if someone can confirm it doesnt break GKE

derekwaynecarr commented:
@pmorie - I addressed nits.  I agree with @erictune that this should get LGTM, and we can do the volume e2e as a follow-up.  I think a googler (@dchen1107 ?) should apply the tag because this previously caused issue with GKE (which I am unable to debug), but I think this should now work without issue in GKE.  Either way, to avoid the revert revert revert mess, it would be good if someone can test that env prior to merge.

pmorie commented:
Sorry, I somehow missed @derekwaynecarr's comment.  Nits have been addressed to satisfaction; I'll let @dchen1107 or another googler apply the tag.

dchen1107 commented:
LGTM

dchen1107 commented:
@andyzheng0831 Can you help us to verify this again? I believe the issue is fixed, but want to double check it against GKE. Thanks!

dchen1107 commented:
@andyzheng0831 Once the test is passed, please apply lgtm label to the pr directly. I did verify it on a single node, but want to double check against GKE environment. Thanks!

andyzheng0831 commented:
I verified this PR in GKE. I first tried to just "git clone" the branch where this PR is in, but it does not have a tag and "git describe" fails. So I had to manually merge the changes in this PR into master branch head for test.

1. GKE cluster creation works;
2. GKE e2e tests fail on one PD case. Later I reran that particular test several times and all passed. So I think it is flaky.

In summary, it should be safe to continue for merge. I will add a lgtm label

dchen1107 commented:
Thanks, @andyzheng0831

roberthbailey commented:
@andyzheng0831 

> I verified this PR in GKE. I first tried to just "git clone" the branch where this PR is in, but it does not have a tag and "git describe" fails. So I had to manually merge the changes in this PR into master branch head for test.

@zmerlynn gave me a nice set of commands that will allow you to easily build/test any pull request:

```
declare -r PULL="${1}"
git fetch upstream pull/${PULL}/head:test_pull_${PULL}
git checkout test_pull_${PULL}
```

I have it in a script `test-pull-request.sh` that I run whenever I need to run someone else's code before it's submitted.

vishh commented:
@derekwaynecarr @dchen1107 Thanks for picking this up and getting it into the release :)
